bindings/java/mongocrypt/src/main/java/com/mongodb/crypt/capi/CipherCallback.java

```diff
@@ -37,7 +43,17 @@ class CipherCallback implements mongocrypt_crypto_fn {
     private final String transformation;
     private final int mode;
 
+    // `cipherPoolMap` is a map from a String transformation (e.g. "AES/CBC/NoPadding") to a pool of Cipher instances.
+    // `cipherPoolMap` is shared by all instances of CipherCallback and is not synchronized.
+    // `cipherPoolMap` is written in a static initializer and later only read.
+    private static final Map<String, CipherPool> cipherPoolMap = new HashMap<>();
```

Comment:
Do we need/want the same treatment for Mac and MessageDigest?

Reply:
I expect no. 

Profiling the benchmark before these changes reported 8.6% samples in `Cipher.getInstance()` and <1% samples in `Mac.getInstance()`:

![Profile before this PR](https://github.com/mongodb/libmongocrypt/assets/992997/7213e658-fc91-4044-bd4c-781e7eab2c4b)


[A patch build](https://spruce.mongodb.com/version/65084f00c9ec444cc5ca395d/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) including a Mac pool showed no significant improvement over these changes:

| 1 threads   | 2 threads   | 8 threads   | 64 threads   |
|-------------|-------------|-------------|--------------|
| 60 ops/sec  | 112 ops/sec | 299 ops/sec | 267 ops/sec  |

`MessageDigest` is only used for SHA-256. SHA-256 is only used to sign KMS messages. Signing KMS messages is only needed to decrypt the Data Encryption Key (DEK) and is infrequent.